boards/xtensa/esp32/CMakeLists.txt

```diff
@@ -60,5 +60,93 @@ if(CONFIG_BOOTLOADER_ESP_IDF)
   board_finalize_runner_args(esp32 "--esp-partition-table-address=0x8000")
 
   board_finalize_runner_args(esp32 "--esp-app-address=0x10000")
+endif()
+
+if(CONFIG_BOOTLOADER_MCUBOOT)
+  include(ExternalProject)
+  include(${CMAKE_CURRENT_LIST_DIR}/dts_utils.cmake)
+  set(mcuboot_dir   ${ZEPHYR_BASE}/../bootloader/mcuboot)
+  set(mcuboot_esp_dir ${mcuboot_dir}/boot/espressif)
+  set(mcuboot_build_dir ${CMAKE_BINARY_DIR}/mcuboot)
+
+  file(COPY ${CMAKE_CURRENT_LIST_DIR}/mcuboot.defconfig DESTINATION ${mcuboot_build_dir})
+
+  get_partition_info("${CMAKE_CURRENT_LIST_DIR}/esp32.dts" "mcuboot" mcub_off
+	  mcub_size)
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_BOOTLOADER_SIZE=${mcub_size}\n")
+
+  get_partition_info("${CMAKE_CURRENT_LIST_DIR}/esp32.dts" "image-0" img_0_off
+	  img_0_size)
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_APPLICATION_PRIMARY_START_ADDRESS=${img_0_off}\n"
+    "CONFIG_ESP_APPLICATION_SIZE=${img_0_size}\n")
+
+  get_partition_info("${CMAKE_CURRENT_LIST_DIR}/esp32.dts" "image-1" img_1_off
+	  img_1_size)
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_APPLICATION_SECONDARY_START_ADDRESS=${img_1_off}\n"
+  )
+
+  get_partition_info("${CMAKE_CURRENT_LIST_DIR}/esp32.dts" "image-scratch" img_scr_off
+	  img_scr_size)
+  file(APPEND ${mcuboot_build_dir}/mcuboot.defconfig
+    "CONFIG_ESP_SCRATCH_OFFSET=${img_scr_off}\n"
+    "CONFIG_ESP_SCRATCH_SIZE=${img_scr_size}\n"
+  )
+
+  add_custom_target(mcub_conf DEPENDS ${mcuboot_build_dir}/mcuboot.defconfig)
+
+  ExternalProject_Add(
+    McuBoot
```

Comment:
> Since this support is for the entire esp32 architecture, I similarly think that putting it in a single board directory looks like the wrong architectural design and this should be factored out so it can be reused by all boards in the arch.

@mbolivar-nordic  I agree with your comment. I have moved bootloader building code to SoC specific CMakeLists.txt, so the change will be applicable to all ESP32 based boards. Please check if the current implementation looks fine.

Reply:
@tejlmand @mbolivar-nordic, was the explanation above enough for you to approve this approach? If not, can you suggest add this into the TSC agenda or dev-review we can talk and move forward on this? Thanks.

Reply:
> ESP32 has a hard dependency on second stage bootloader for setting up XiP. None of the samples would work without second stage bootloader.

Might be.
But how is that different from any other sample where MCUBoot is enabled ?
If a bootloader is enabled, you must flash that bootloader before the sample will work.

Our guide clearly states that a bootloader must be flashed independently:
https://docs.zephyrproject.org/latest/guides/device_mgmt/dfu.html#bootloaders
> You need to build and flash MCUboot itself on your device

Zephyr has no dependency to the build of the bootloader in this case.
If I disable the `ExternalProject` code, then Zephyr builds just fine:
```
[140/140] Linking C executable zephyr/zephyr.elf
Memory region         Used Size  Region Size  %age Used
     mcuboot_hdr:          32 B         32 B    100.00%
        metadata:          32 B         32 B    100.00%
             ROM:       74996 B    4194240 B      1.79%
     iram0_0_seg:       11460 B       128 KB      8.74%
     irom0_0_seg:        9524 B      3264 KB      0.28%
     dram0_0_seg:        1576 B     180736 B      0.87%
     dram0_1_seg:         53 KB     113840 B     47.67%
     drom0_0_seg:        1832 B    4194240 B      0.04%
    rtc_iram_seg:          0 GB         8 KB      0.00%
    rtc_slow_seg:          0 GB         4 KB      0.00%
        IDT_LIST:          0 GB         8 KB      0.00%
```

That is very different from the TF-M case.
The TF-M build produces libraries and object files that is needed by the Zephyr build system in order to link Zephyr itself:
```
arm-zephyr-eabi-gcc
  ...
  -o zephyr/zephyr_prebuilt.elf
  ...
  tfm/platform/libplatform_ns.a  tfm/app/libtfm_api_ns.a  tfm/secure_fw/s_veneers.o
  ...
```
if I try to disable `TF-M` external project on such build, then i'll get:
```
$ ninja -Cbuild
ninja: Entering directory `build'
ninja: error: 'tfm/app/libtfm_api_ns.a', needed by 'zephyr/zephyr_prebuilt.elf', missing and no known rule to make it
```
or
```
$ ninja -Cbuild
ninja: Entering directory `build'
ninja: error: 'tfm/secure_fw/s_veneers.o', needed by 'zephyr/zephyr_prebuilt.elf', missing and no known rule to make it
```

It has been discussed before at TSC that Zephyr should not do multi-image builds in it's CMake build system just for convenience of the user.
Multi image build should be kept outside Zephyr build system.
The reason TF-M is special is because the built output from TF-M is needed for the linking stage of Zephyr itself.

Looking at history I notice that Espressif has been doing this for long time starting here:
https://github.com/zephyrproject-rtos/zephyr/pull/8383

Therefore I think this should be brought up again and discussed at the TSC.
And if approved there, note down under which circumstances the Zephyr build system should allow this approach.

I'm not against introducing multi-image builds and flashing for the user-convenience, but I would like TSC approval on this with clear guidelines on when / how this should be handled. Especially because this question has be raised before.

Reply:
Closing this thread as this has been changed and no longer make sense.